homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,356 @@
+"""Provides functionality to interact with humidifier devices."""
+from abc import abstractmethod
+from datetime import timedelta
+import logging
+from typing import Any, Dict, List, Optional
+
+import voluptuous as vol
+
+from homeassistant.const import SERVICE_TURN_OFF, SERVICE_TURN_ON, STATE_OFF, STATE_ON
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+    make_entity_service_schema,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType, ServiceDataType
+
+from .const import (
+    ATTR_AUX_HEAT,
+    ATTR_CURRENT_HUMIDITY,
+    ATTR_CURRENT_TEMPERATURE,
+    ATTR_FAN_MODE,
+    ATTR_FAN_MODES,
+    ATTR_HUMIDIFIER_ACTION,
+    ATTR_HUMIDITY,
+    ATTR_MAX_HUMIDITY,
+    ATTR_MIN_HUMIDITY,
+    ATTR_OPERATION_MODE,
+    ATTR_OPERATION_MODES,
+    ATTR_PRESET_MODE,
+    ATTR_PRESET_MODES,
+    ATTR_WATER_LEVEL,
+    DEFAULT_MAX_HUMIDITY,
+    DEFAULT_MIN_HUMIDITY,
+    DOMAIN,
+    OPERATION_MODE_DRY,
+    OPERATION_MODE_HUMIDIFY,
+    OPERATION_MODE_HUMIDIFY_DRY,
+    OPERATION_MODE_OFF,
+    OPERATION_MODES,
+    SERVICE_SET_AUX_HEAT,
+    SERVICE_SET_FAN_MODE,
+    SERVICE_SET_HUMIDITY,
+    SERVICE_SET_OPERATION_MODE,
+    SERVICE_SET_PRESET_MODE,
+    SUPPORT_AUX_HEAT,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TEMPERATURE,
+    SUPPORT_WATER_LEVEL,
+)
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up humidifier devices."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(SERVICE_TURN_ON, {}, "async_turn_on")
+    component.async_register_entity_service(SERVICE_TURN_OFF, {}, "async_turn_off")
+    component.async_register_entity_service(
+        SERVICE_SET_FAN_MODE,
+        make_entity_service_schema({vol.Required(ATTR_FAN_MODE): cv.string}),
+        "async_set_fan_mode",
+        [SUPPORT_FAN_MODE],
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_OPERATION_MODE,
+        make_entity_service_schema(
+            {vol.Required(ATTR_OPERATION_MODE): vol.In(OPERATION_MODES)}
+        ),
+        "async_set_operation_mode",
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_PRESET_MODE,
+        make_entity_service_schema({vol.Required(ATTR_PRESET_MODE): cv.string}),
+        "async_set_preset_mode",
+        [SUPPORT_PRESET_MODE],
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_AUX_HEAT,
+        make_entity_service_schema({vol.Required(ATTR_AUX_HEAT): cv.boolean}),
```

Comment:
aux heat? This is a humidifier.

Reply:
Some humidifiers feature 'heated mist'.

Reply:
Then why name it "aux"?

Reply:
Probably because English is not my native language:) I understand the word `auxiliary` as `additional`, and because heating the mist is not a main feature for a humidifier, but `additional`, I thought the term would do.
What should I rename it to? The user manuals call it 'warm mist' or 'heated mist', but since I was hoping this integration would work for dehumidifiers as well, I decided not to use the word 'mist'.